src/SerialIO.c

```diff
@@ -0,0 +1,58 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "SerialIO.h"
+#include <stdio.h>
+#include "Serial.h"
+
+static int
+Print(uint32_t aLength, void* aBuffer)
+{
+  IPCMessageQueue* queue = GetSerialOutQueue();
+  if (!queue) {
+    return -1;
+  }
+
+  IPCMessage msg;
+  int res = IPCMessageInit(&msg);
+  if (res < 0) {
+    return -1;
+  }
+  res = IPCMessageProduce(&msg, aLength, aBuffer);
+  if (res < 0) {
+    return -1;
+  }
+  res = IPCMessageQueueConsume(queue, &msg);
+  if (res < 0) {
+    return -1;
+  }
+  return IPCMessageWaitForConsumption(&msg);
+}
+
+/*
+ * Libc-like interfaces for easy usage.
+ */
+
+int
+print(const char* fmt, ...)
+{
+  va_list ap;
+
+  va_start(ap, fmt);
+  int res = vprint(fmt, ap);
+  va_end(ap);
+
+  return res;
+}
+
+int
+vprint(const char* fmt, va_list ap)
+{
+  char buf[128];
+  int res = vsnprintf(buf, sizeof(buf), fmt, ap);
+  if (res < 0) {
+    return -1;
+  }
+  return Print(sizeof(buf), buf);
```

Comment:
Yeah - the StrXPrintf code doesn't have any concurrency in it, since it doesn't really make sense to have any in it. Concurrency needs to be dealt with outside of the formatting.

Some users might want to acquire a lock, call StrXPrintf and pass in a function which writes to the serial port. Some users might just want to call StrXPrintf and have the output function do the necessary mutual exclusion (I tend to prefer the former). Functions like sprintf don't do any concurrency control either.

The big advantage of using StrXPrintf is that you don't need to create a temporary buffer to format the output into. You just provide a function which takes characters and outputs them to your LCD or serial port, or whatever.

The license at the top of StrPrintf.c already states that the code is considered public domain, so you can do with it as you wish.